learningequality/studio
# Edit modal - add "Category" field to "Basic information" section

## Summary
### Description of the change(s) you made

- Add category field to edit modal
- Add tests for category field component
- Commit history will be cleaned, and needs to be rebased after #3370 is merged

### Manual verification steps performed
1. Checked to see if behavior of the checkboxes in the dropdown and the chips work appropriately
2. Checked for data to be sent correctly to the backend
3. Checked to see if tests passed

### Screenshots
|Chip and dropdown|
|--|
|<img width="973" alt="Screen Shot 2022-04-25 at 9 01 30 PM" src="https://user-images.githubusercontent.com/13563002/165218847-a1be2e0d-a77f-4e09-9584-a2757da71162.png">|

|Chip overflow|
|--|
|<img width="973" alt="Screen Shot 2022-04-25 at 9 02 03 PM" src="https://user-images.githubusercontent.com/13563002/165218694-96883e6b-c225-4640-8482-bcbe1d972cf0.png">|

### Does this introduce any tech-debt items?
The code is pretty spaghetti-ish and could use some refactoring, particularly around the behaviors of the dropdown. It should, however, currently satisfy what is needed for an MVP. We can file follow-up issues to fix. This is a continuation of this issue https://github.com/learningequality/studio/issues/3205
___
## Reviewer guidance
### How can a reviewer test these changes?
- Any feedback on the tests would be very helpful. I'm not sure the best way to test the data being passed in (yet)
- Manual QA will be very important! I've tried to capture as many of the interactions as I could think of below

### Are there any risky areas that deserve extra testing?
It is very important to refer to the Gherkin stories and the Figma designs. There are additional scenarios that should be tested that are not described in the scenarios or in the Figma designs, but should be noted (I may also be missing some):

**There are grandparents, parents, and children items**
- Check any item with no children
  - dropdown: only that item should be checked
  - chips: only that item should appear
- Check a parent item with no children
  - dropdown: that item and the grandparent should be checked
  - chips: only that item should appear
- Check any item with children
  - dropdown: that item and its parents should be checked
  - chips: only that item should appear
- Check any item, and then check its child
  - dropdown: that item and its parents should be checked
  - chips: that item should be replaced with its child

- Uncheck any child with no siblings and its parent has no siblings checked
  - dropdown: that item and all its parents should be unchecked
  - chips: remove it from chips
- Uncheck any child with no siblings but its parent has siblings checked
  - dropdown: that item and its immediate parent should be unchecked
  - chips: remove it from chips
- Uncheck any child with siblings and its parent has no siblings checked
  - dropdown: only that item should be unchecked
  - chips: remove that item from chips
- Uncheck any child with siblings and its parent has siblings checked
  - dropdown: only that item should be unchecked
  - chips: remove that item from chips
- Uncheck a parent with checked children
  - dropdown: remove that parent and its checked children
  - chips: remove the child chip

## References
- [Figma design for this feature](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=1764%3A88434)
- [Gherkin stories for this feature](https://www.notion.so/learningequality/Studio-gherkin-stories-6bb841246d4e4afd958f074226fa8f19)
- Addresses https://github.com/learningequality/studio/issues/3205

----

### Contributor's Checklist
PR process:

Studio-specifc:

Testing:
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
